Auto download C Driver (#967)
mongodb/mongo-cxx-driver: CMakeLists.txt

  message(WARNING "Unknown compiler... recklessly proceeding without a version check")
endif()

set(LIBMONGOC_REQUIRED_VERSION 1.22.1) # TODO: update to 1.24.0 once released.
set(LIBMONGOC_REQUIRED_ABI_VERSION 1.0)

set(NEED_DOWNLOAD_C_DRIVER false)
if(TARGET mongoc_shared OR TARGET mongoc_static)
  # If these targets exist, then libmongoc has already been included as a project
  # sub-directory
  message (STATUS "Found libmongoc targets declared in current build scope; version not checked")

  if(NOT MONGOCXX_LINK_WITH_STATIC_MONGOC)
    set(libmongoc_target mongoc_shared)
  else()
    set(libmongoc_target mongoc_static)
  endif()

  if(MONGOCXX_BUILD_STATIC)
    set(mongocxx_pkg_dep "find_dependency(mongoc-1.0 REQUIRED)")
  endif()
else()
    find_package(mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION} ${LIBMONGOC_REQUIRED_VERSION} QUIET)
    if(mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION}_FOUND)

>         message(STATUS "Already found libmongoc, don't need to download it.")

Reviewer: ```suggestion
        message(STATUS "Found libmongoc ${mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION}_VERSION}, don't need to download it.")
```

Suggest printing found version.

    else()
        set(NEED_DOWNLOAD_C_DRIVER true CACHE INTERNAL "")
    endif()
endif()

if(NEED_DOWNLOAD_C_DRIVER)
    message(STATUS "No Mongo C Driver path provided via CMAKE_PREFIX_PATH, will download a release from the internet.")
    include(FetchContent)
    include(ExternalProject)

    # Declare mongo-c-driver as a dependency
    FetchContent_Declare(
      mongo-c-driver
      GIT_REPOSITORY https://github.com/mongodb/mongo-c-driver.git
      GIT_TAG b9ca36ff8c64f12330326f080c725c5502294834
    )

>     FetchContent_MakeAvailable(mongo-c-driver)

Reviewer: The current changes also install the C driver:

```
cmake -S. -Bcmake-build -DCMAKE_CXX_STANDARD=17
cmake --build cmake-build --target install
# Results in installing C driver files. Example:
# /Users/kevin.albertson/review/mongo-cxx-driver-967/cmake-build/install/lib/libmongoc-1.0.0.dylib
```

Consumers of the C++ driver may not need a separate install of the C driver.

Consider disabling install of the C driver, and static linking to the C driver.

Author: Sounds good!

    # Set the installation directory for mongo-c-driver

>     set(MONGOC_INSTALL_DIR ${CMAKE_BINARY_DIR}/mongo-c-driver-install)

Author: Ah, it was used as the location where the C library would be installed to, but then it was removed by this suggestion: 
![image](https://github.com/mongodb/mongo-cxx-driver/assets/13429210/1d06f173-2b28-44b9-a3df-d1c74606139d)

  message(WARNING "Unknown compiler... recklessly proceeding without a version check")
endif()

set(LIBMONGOC_REQUIRED_VERSION 1.22.1) # TODO: update to 1.24.0 once released.
set(LIBMONGOC_REQUIRED_ABI_VERSION 1.0)

> set(NEED_DOWNLOAD_C_DRIVER false CACHE INTERNAL "")

Reviewer 2: Use a regular variable instead of `CACHE`, since this is unconditionally assigned.

Author: Ok, this seems to be resolved with this commit: https://github.com/mongodb/mongo-cxx-driver/pull/967/commits/fec94f4ae58c790d2edc67245b55d6f2f429a811

if(TARGET mongoc_shared OR TARGET mongoc_static)
  # If these targets exist, then libmongoc has already been included as a project
  # sub-directory

>   message ("found libmongoc targets declared in current build scope; version not checked")

Reviewer 2: Always use a `message()` mode indicator

```suggestion
  message (STATUS "Found libmongoc targets declared in current build scope; version not checked")
```

  if(NOT MONGOCXX_LINK_WITH_STATIC_MONGOC)
    set(libmongoc_target mongoc_shared)
  else()
    set(libmongoc_target mongoc_static)
  endif()

  if(MONGOCXX_BUILD_STATIC)
    set(mongocxx_pkg_dep "find_dependency(mongoc-1.0 REQUIRED)")
  endif()
else()
    find_package(mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION} ${LIBMONGOC_REQUIRED_VERSION} QUIET)
    if(mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION}_FOUND)

>         message("already found libmongoc, don't need to download it.")

Reviewer 2: ```suggestion
        message(STATUS "Already found libmongoc, don't need to download it.")
```

    else()
        set(NEED_DOWNLOAD_C_DRIVER true CACHE INTERNAL "")
    endif()
endif()

> if (${NEED_DOWNLOAD_C_DRIVER})

Reviewer 2: Boolean variables need not be dereferenced. This is one of the very weird quirks in CMake syntax: Refer to [Variable Expansion](https://cmake.org/cmake/help/latest/command/if.html#variable-expansion) on the page for `if()`.

```suggestion
if (NEED_DOWNLOAD_C_DRIVER)
```

>     message("No Mongo C Driver path provided via CMAKE_PREFIX_PATH, will download a release from the internet.")

Reviewer 2: ```suggestion
    message(STATUS "No Mongo C Driver path provided via CMAKE_PREFIX_PATH, will download a release from the internet.")
```

    include(FetchContent)
    include(ExternalProject)

    # Declare mongo-c-driver as a dependency
    FetchContent_Declare(
      mongo-c-driver
      GIT_REPOSITORY https://github.com/mongodb/mongo-c-driver.git
      GIT_TAG b9ca36ff8c64f12330326f080c725c5502294834
    )

    FetchContent_MakeAvailable(mongo-c-driver)

    # Set the installation directory for mongo-c-driver
    set(MONGOC_INSTALL_DIR ${CMAKE_BINARY_DIR}/mongo-c-driver-install)

    # Build mongo-c-driver using ExternalProject
    ExternalProject_Add(mongo-c-driver-build
        SOURCE_DIR ${mongo-c-driver_SOURCE_DIR}
        BINARY_DIR ${CMAKE_BINARY_DIR}/mongo-c-driver-build
        INSTALL_DIR ${MONGOC_INSTALL_DIR}
        CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${MONGOC_INSTALL_DIR}

>     )

Reviewer 2: This can be removed, as `FetchContent_MakeAvailable` will automatically call `add_subdirectory()` on the downloaded tree, making the libmongoc and libbson targets available in the rest of the build. This also alleviates the need to manually create the build dependencies and update the include directories of downstream targets.

    else()
        set(NEED_DOWNLOAD_C_DRIVER true CACHE INTERNAL "")
    endif()
endif()

if(NEED_DOWNLOAD_C_DRIVER)
    message(STATUS "No Mongo C Driver path provided via CMAKE_PREFIX_PATH, will download C driver version ${LIBMONGOC_DOWNLOAD_VERSION} from the internet.")
    include(FetchContent)
    include(ExternalProject)

    set(BUILD_SHARED_LIBS_WITH_STATIC_MONGOC ON CACHE INTERNAL "")

    # Declare mongo-c-driver as a dependency
    FetchContent_Declare(
      mongo-c-driver
      GIT_REPOSITORY https://github.com/mongodb/mongo-c-driver.git
      GIT_TAG ${LIBMONGOC_DOWNLOAD_VERSION}
    )

    FetchContent_GetProperties(mongo-c-driver)
    if(NOT mongo-c-driver_POPULATED)
      FetchContent_Populate(mongo-c-driver)
      add_subdirectory(${mongo-c-driver_SOURCE_DIR} ${mongo-c-driver_BINARY_DIR} EXCLUDE_FROM_ALL)
    endif()

    # Set the installation directory for mongo-c-driver
    set(MONGOC_INSTALL_DIR ${CMAKE_BINARY_DIR}/mongo-c-driver-install)

    # These scripts are not needed, and cause the installation to freeze.
    file(REMOVE ${mongo-c-driver_SOURCE_DIR}/build/generate-uninstall.sh)

>     file(REMOVE ${mongo-c-driver_SOURCE_DIR}/build/generate-uninstall.cmd)

Reviewer 3: I wonder if this might be able to go away if the C driver isn't being installed.

Author: Sweet! You're right!

  message(WARNING "Unknown compiler... recklessly proceeding without a version check")
endif()

set(LIBMONGOC_REQUIRED_VERSION 1.24.0)

> set(LIBMONGOC_DOWNLOAD_VERSION ba5ab6de26a874d33b0abc3d2b46961a69380e7a)

Reviewer: ```suggestion
set(LIBMONGOC_DOWNLOAD_VERSION ba5ab6de26a874d33b0abc3d2b46961a69380e7a) # TODO: update to 1.25.0 once C driver 1.25.0 is released.
```

  if(NOT MONGOCXX_LINK_WITH_STATIC_MONGOC)
    set(libmongoc_target mongoc_shared)
  else()
    set(libmongoc_target mongoc_static)
  endif()

  if(MONGOCXX_BUILD_STATIC)
    set(mongocxx_pkg_dep "find_dependency(mongoc-1.0 REQUIRED)")
  endif()
else()
    find_package(mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION} ${LIBMONGOC_REQUIRED_VERSION} QUIET)
    if(mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION}_FOUND)
        message(STATUS "Found libmongoc ${mongoc-${LIBMONGOC_REQUIRED_ABI_VERSION}_VERSION}, don't need to download it.")
    else()
        set(NEED_DOWNLOAD_C_DRIVER true CACHE INTERNAL "")
    endif()
endif()

if(NEED_DOWNLOAD_C_DRIVER)
    message(STATUS "No Mongo C Driver path provided via CMAKE_PREFIX_PATH, will download C driver version ${LIBMONGOC_DOWNLOAD_VERSION} from the internet.")
    include(FetchContent)
    include(ExternalProject)

    # Declare mongo-c-driver as a dependency
    FetchContent_Declare(
      mongo-c-driver
      GIT_REPOSITORY https://github.com/mongodb/mongo-c-driver.git
      GIT_TAG ${LIBMONGOC_DOWNLOAD_VERSION}
    )

    if(NOT CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")

>         set(CMAKE_C_FLAGS -Wno-error)

Author: This was added because there are some warnings that are triggered by the C driver dependency, which it would seem to me that warnings from the C driver should not cause build failures for the C++ driver.

    endif()

    FetchContent_GetProperties(mongo-c-driver)
    if(NOT mongo-c-driver_POPULATED)
        set(OLD_ENABLE_TESTS ${ENABLE_TESTS})
        set(OLD_BUILD_TESTING ${BUILD_TESTING})
        set(OLD_CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
        set(OLD_CMAKE_C_FLAGS ${CMAKE_C_FLAGS})

        set(ENABLE_EXTRA_ALIGNMENT OFF)

>         add_compile_definitions(ENABLE_EXTRA_ALIGNMENT=OFF)

Reviewer: ```suggestion
```
Remove? I expect `set(ENABLE_EXTRA_ALIGNMENT OFF)` disables extra alignment in the included C driver.

Author: In a prior meeting that we had, you suggested disabling extra alignment as a way to fix the UBSAN issues that were caused by extra alignment.

Author: I stand corrected, it appears that `set(ENABLE_EXTRA_ALIGNMENT OFF)` alone is sufficient to fix this. Thank you

[Patch](https://spruce.mongodb.com/version/64c29d8e56234360d15ae2ce/tasks?page=0&sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC&variant=%5Eubuntu1804-debug-ubsan-50%24)

        FetchContent_Populate(mongo-c-driver)
        # Set ENABLE_TESTS to OFF to disable the test-libmongoc target in the C driver.
        # This prevents the LoadTests.cmake script from attempting to execute test-libmongoc.
        # test-libmongoc is not built with the "all" target.
        # Attempting to execute test-libmongoc results in an error: "Unable to find executable: NOT_FOUND"
        set(ENABLE_TESTS OFF)
        set(BUILD_TESTING OFF)

>         string(REPLACE " -Werror" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")

Reviewer: Is removing `-Werror` from `CMAKE_CXX_FLAGS` necessary? I expect not for including the C driver.

Author: Without removing `-Werror` from the CXX flags here, we get the following error when building `mongocxx_mocked`:

```
[2023/08/01 13:15:28.405] [ 56%] Building CXX object src/mongocxx/CMakeFiles/mongocxx_mocked.dir/client_encryption.cpp.o
[2023/08/01 13:15:28.574] In file included from /data/mci/8c99f2b8789a9a955d3047097dbfa1be/mongo-cxx-driver/build/_deps/mongo-c-driver-src/src/libbson/src/bson/bson-types.h:30:0,
[2023/08/01 13:15:28.574]                  from /data/mci/8c99f2b8789a9a955d3047097dbfa1be/mongo-cxx-driver/build/_deps/mongo-c-driver-src/src/libbson/src/bson/bson-context.h:25,
[2023/08/01 13:15:28.574]                  from /data/mci/8c99f2b8789a9a955d3047097dbfa1be/mongo-cxx-driver/build/_deps/mongo-c-driver-src/src/libbson/src/bson/bson.h:32,
[2023/08/01 13:15:28.574]                  from /data/mci/8c99f2b8789a9a955d3047097dbfa1be/mongo-cxx-driver/build/_deps/mongo-c-driver-src/src/cpp-check.cpp:5:
[2023/08/01 13:15:28.574] /data/mci/8c99f2b8789a9a955d3047097dbfa1be/mongo-cxx-driver/build/_deps/mongo-c-driver-src/src/libbson/src/bson/bson-endian.h: In function 'uint16_t __bson_uint16_swap_slow(uint16_t)':
[2023/08/01 13:15:28.574] /data/mci/8c99f2b8789a9a955d3047097dbfa1be/mongo-cxx-driver/build/_deps/mongo-c-driver-src/src/libbson/src/bson/bson-endian.h:136:73: error: conversion to 'uint16_t {aka short unsigned int}' from 'int' may alter its value [-Werror=conversion]
[2023/08/01 13:15:28.574]     return (uint16_t) ((v & 0x00FF) << 8) | (uint16_t) ((v & 0xFF00) >> 8);
```

Logs found [here](https://spruce.mongodb.com/task/cxx_driver_ubuntu1604_debug_std_experimental_compile_and_test_with_shared_libs_patch_cd9542cbf0186935f055e3e03469764033f70457_64c903293627e0e45398cfdb_23_08_01_13_05_46/logs?execution=0)

        string(REPLACE " -Werror" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
        add_subdirectory(${mongo-c-driver_SOURCE_DIR} ${mongo-c-driver_BINARY_DIR})
        set(CMAKE_CXX_FLAGS ${OLD_CMAKE_CXX_FLAGS})
        set(CMAKE_C_FLAGS ${OLD_CMAKE_C_FLAGS})
        set(ENABLE_TESTS OLD_ENABLE_TESTS)
        set(BUILD_TESTING OLD_BUILD_TESTING)

>     endif()

Reviewer: ```suggestion
    endif()
    message(STATUS "Download and configure C driver version ${LIBMONGOC_DOWNLOAD_VERSION} ... end")
```

Suggest adding a `begin` and `end` message to make it easier to identify the C driver cmake output in logs.

      ${ALL_DIST} ${dist_generated_depends}
)

> if(NOT (TARGET distcheck))

Reviewer: ```suggestion
if(NOT (TARGET dist OR TARGET distcheck))
```

Since this block defines both `dist` and `distcheck`.
